File: ruby/lib/coding_interview/application/usecase/portfolio/get_latest_portfolio_usecase.rb
module CodingInterview
  module Application
    module Usecase
      module Portfolio
        GetLatestPortfolioItemOutput = Struct.new(:symbol, :rate)
        GetLatestPortfolioUsecaseOutput = Struct.new(:items)

        class GetLatestPortfolioUsecase
          def initialize(portfolio_repository)
            @portfolio_repository = portfolio_repository
          end

          def run
            p = @portfolio_repository.get
            GetLatestPortfolioUsecaseOutput.new(
              p.items.map { |i| GetLatestPortfolioItemOutput.new(i.symbol, i.rate) }
            )
          end
        end
      end
    end
  end
end

File: python/src/coding_interview/application/usecase/portfolio/get_latest_portfolio_usecase.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from coding_interview.application.repository.portfolio_repository import PortfolioRepository
from coding_interview.domain.stock_symbol import StockSymbol


@dataclass(frozen=True)
class GetLatestPortfolioItemOutput:
    symbol: StockSymbol
    rate: Decimal


@dataclass(frozen=True)
class GetLatestPortfolioUsecaseOutput:
    items: tuple[GetLatestPortfolioItemOutput, ...]


class GetLatestPortfolioUsecase:
    def __init__(self, portfolio_repository: PortfolioRepository) -> None:
        self._portfolio_repository = portfolio_repository

    def run(self) -> GetLatestPortfolioUsecaseOutput:
        portfolio = self._portfolio_repository.get()
        items = tuple(
            GetLatestPortfolioItemOutput(symbol=item.symbol, rate=item.rate)
            for item in portfolio.items
        )
        return GetLatestPortfolioUsecaseOutput(items=items)
